docs/RFCS/monitoring_queries_and_sessions.md

```diff
@@ -0,0 +1,219 @@
+both the active query reference and the `parallelizeQueue` will be required to know all queries
+in flight.
+
+Write locks on this struct will be acquired when:
+- A new session is created (all fields are populated)
+- The application name is reset (in `resetApplicationName()`)
+- A query begins or ends execution (to updated the active query reference / parallelize queue)
+- txnState.txn is updated in `executor.execRequest`. 
+
+## Node-local session registry
+
+The session registry will be an in-memory slice that stores references to all
+session objects on that node. Addition and deletion of new
+sessions to this slice would require acquiring a mutex lock.
+
+This addition/deletion into the slice can be made in `NewSession(...)` and `Session.Finish(...)`
+in `pkg/sql/session.go`.
+
+## New Query struct
+
+A new struct, `Query`, would be made that contains the below fields:
+
+- Query string (output of `planToString()`)
+- Start timestamp
+- Reference to root `planNode` (which it could effectively replace)
+- Eventually: any contexts/object references useful for cancellation 
+
+The parallelizeQueue will be updated to take Query objects instead of planNodes. Non-parallelized
+queries will be directly referenced via a pointer in `Session`.
+
+## RPC endpoint in status server to return local queries and sessions
+
+Two new RPC endpoints would be made: `LocalSessions` and `LocalQueries`. These
+endpoints would return the node's local store of active sessions and queries respectively.
+
+These would be added to the status RPC server.
+
+## SQL statements to retrieve cluster-wide queries and sessions
+
+For sessions, a new SQL statement will be added to the grammar: `SHOW SESSIONS [GLOBAL|LOCAL]`.
+`SHOW SESSIONS LOCAL` will return the gateway node's local list of sessions only, while
+`SHOW SESSIONS GLOBAL` will return the entire cluster's list of sessions. Default is GLOBAL
+if unspecified.
+ 
+The `GLOBAL` call will involve looping through all nodes and issue RPC calls to `LocalSessions`
+with timeouts. The results of all the calls that do succeed, will be returned as a table.
+`RaftDebug` in `pkg/server/status.go` is an example of a procedure that does a similar aggregation
+of results across all known nodes in the cluster.
+
+Example usage:
+
+```
+
+root@:26257/> SHOW SESSIONS
+
++---------+----------+------------------+------------------+------------------------------------------------------+---------------+--------------------+-----------+
+| node_id | username | client_address   | application_name | active_queries                                       | session_start | oldest_query_start | kv_txn_id |
++---------+----------+------------------+------------------+------------------------------------------------------+---------------+--------------------+-----------+
+| 1       |  root    | 192.168.1.5:8080 | test             | SELECT * FROM users; INSERT INTO test VALUES (...);  | 1234567890    | 1234567891         | 1e5f6bac  |
```

Comment:
I think there is a way to anonymize these queries such that you show no private data in the response